core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -129,15 +159,15 @@ private[kafka] object LogValidator extends Logging {
         }
       }
 
-      if (batch.hasProducerId && batch.baseSequence < 0) {
+      if (batch.isControlBatch) {
         brokerTopicStats.allTopicsStats.invalidOffsetOrSequenceRecordsPerSec.mark()
-        throw new InvalidRecordException(s"Invalid sequence number ${batch.baseSequence} in record batch " +
-          s"with producerId ${batch.producerId} in topic partition $topicPartition.")
+        throw new InvalidRecordException(s"Clients are not allowed to write control records in topic partition $topicPartition.")
```

Comment:
I think I ran into this by mistake while testing. I had sent a control record with `origin == Client` which resulted in the `Invalid sequence number` error below instead of the more specific error here. So I reversed the order.